packages/toolkit/src/codewhisperer/service/transformByQHandler.ts

```diff
@@ -78,89 +77,123 @@ export async function getOpenProjects() {
-            errorReason = 'JavapSpawnError'
+        throw new ToolkitError('No Java projects found', { code: 'CouldNotFindJavaProject' })
+    }
+    const mavenJavaProjects = []
+    let containsGradle = false
+    for (const project of javaProjects) {
+        const projectPath = project.description
+        const buildSystem = await checkBuildSystem(projectPath!)
+        if (buildSystem === BuildSystem.Maven) {
+            mavenJavaProjects.push(project)
+        } else if (buildSystem === BuildSystem.Gradle) {
+            containsGradle = true
         }
-        if (spawnResult.error) {
-            const errorCode = (spawnResult.error as any).code ?? 'UNKNOWN'
-            errorReason += `-${errorCode}`
-        }
-        telemetry.codeTransform_isDoubleClickedToTriggerInvalidProject.emit({
-            codeTransformSessionId: codeTransformTelemetryState.getSessionId(),
-            codeTransformPreValidationError: 'NoJavaProject',
-            codeTransformRuntimeError: errorReason,
-            result: MetadataResult.Fail,
-            reason: 'CannotDetermineJavaVersion',
-        })
-        throw new ToolkitError('Unable to determine Java version', {
-            code: 'CannotDetermineJavaVersion',
-        })
     }
-    const majorVersionIndex = spawnResult.stdout.indexOf('major version: ')
-    const javaVersion = spawnResult.stdout.slice(majorVersionIndex + 15, majorVersionIndex + 17).trim()
-    if (javaVersion === CodeWhispererConstants.JDK8VersionNumber) {
-        transformByQState.setSourceJDKVersionToJDK8()
-    } else if (javaVersion === CodeWhispererConstants.JDK11VersionNumber) {
-        transformByQState.setSourceJDKVersionToJDK11()
-    } else {
-        void vscode.window.showErrorMessage(CodeWhispererConstants.noSupportedJavaProjectsFoundMessage)
+    if (mavenJavaProjects.length === 0) {
+        void vscode.window.showErrorMessage(CodeWhispererConstants.noPomXmlFoundMessage)
         telemetry.codeTransform_isDoubleClickedToTriggerInvalidProject.emit({
             codeTransformSessionId: codeTransformTelemetryState.getSessionId(),
-            codeTransformPreValidationError: 'UnsupportedJavaVersion',
+            codeTransformPreValidationError: 'NonMavenProject',
             result: MetadataResult.Fail,
-            reason: javapOutputToTelemetryValue(javaVersion),
+            reason: containsGradle ? 'NoPomFileFoundButGradleFileFound' : 'NoPomFileFound',
         })
-        throw new ToolkitError('Project selected is not Java 8 or Java 11', { code: 'UnsupportedJavaVersion' })
+        throw new ToolkitError('No valid Maven build file found', { code: 'CouldNotFindPomXml' })
+    }
+
+    /*
+     * these projects we know must contain a pom.xml and a .java file
+     * here we try to get the Java version of each project so that we
+     * can pre-select a default version in the QuickPick for them
+     */
+    const projectsValidToTransform = new Map<vscode.QuickPickItem, JDKVersion | undefined>()
```

Comment:
Can we make this a function call like `const projectsValidToTransform = getValidProjectsToTransform()` and just take the following 50 lines or so and make it a function.

Reply:
Blocker

Reply:
Fixed